src/query_filter.rs

```diff
@@ -0,0 +1,343 @@
+
+    unsafe fn init_fetch<'w>(
+        world: &'w World,
+        _state: &Self::State,
+        last_run: Tick,
+        this_run: Tick,
+    ) -> Self::Fetch<'w> {
+        Self::Fetch::<'w> {
+            registry: world.resource(),
+            storage: ChangeDetectionStorage::Uninit,
+            sparse_sets: &world.storages().sparse_sets,
+            last_run,
+            this_run,
+        }
+    }
+
+    unsafe fn clone_fetch<'w>(fetch: &Self::Fetch<'w>) -> Self::Fetch<'w> {
+        Self::Fetch {
+            registry: fetch.registry,
+            storage: match fetch.storage {
+                ChangeDetectionStorage::Uninit => ChangeDetectionStorage::Uninit,
+                ChangeDetectionStorage::Table { ticks } => ChangeDetectionStorage::Table { ticks },
+                ChangeDetectionStorage::SparseSet { components } => {
+                    ChangeDetectionStorage::SparseSet { components }
+                }
+            },
+            sparse_sets: fetch.sparse_sets,
+            last_run: fetch.last_run,
+            this_run: fetch.this_run,
+        }
+    }
+
+    // This will always be false for us, as we (so far) do not know at compile time whether the
+    // components our trait has been impl'd for are stored in table or in sparse set
+    const IS_DENSE: bool = false;
+    const IS_ARCHETYPAL: bool = false;
+
+    #[inline]
+    unsafe fn set_archetype<'w>(
+        fetch: &mut Self::Fetch<'w>,
+        state: &Self::State,
+        _archetype: &'w Archetype,
+        table: &'w Table,
+    ) {
+        // Search for a registered trait impl that is present in the archetype.
+        // We check the table components first since it is faster to retrieve data of this type.
+        for &component in &*state.components {
+            if let Some(column) = table.get_column(component) {
+                fetch.storage = ChangeDetectionStorage::Table {
+                    ticks: ThinSlicePtr::from(column.get_added_ticks_slice()),
+                };
+                return;
+            }
+        }
+        for &component in &*state.components {
+            if let Some(components) = fetch.sparse_sets.get(component) {
+                fetch.storage = ChangeDetectionStorage::SparseSet { components };
+                return;
+            }
+        }
+        // At least one of the components must be present in the table/sparse set.
+        debug_unreachable()
+    }
+
+    #[inline]
+    unsafe fn set_table<'w>(_fetch: &mut Self::Fetch<'w>, _state: &Self::State, _table: &'w Table) {
+        // only gets called if IS_DENSE == true, which does not hold for us
+        debug_unreachable()
+    }
+
+    #[inline(always)]
+    unsafe fn fetch<'w>(
+        fetch: &mut Self::Fetch<'w>,
+        entity: Entity,
+        table_row: TableRow,
+    ) -> Self::Item<'w> {
+        let ticks_ptr = match fetch.storage {
+            ChangeDetectionStorage::Uninit => {
+                // set_archetype must have been called already
```

Comment:
I noticed that in `lib.rs` we have "registry sealing" tracking whether a registry is sealed or open: https://github.com/JoJoJet/bevy-trait-query/blob/2071f9c948358f0959b538ecf54f3a54cc6de923/src/lib.rs#L286

If the seal check fails, a panic is issued: <https://github.com/JoJoJet/bevy-trait-query/blob/main/src/lib.rs#L311>

I am wondering: is it worth putting a similar panic-check here that enforces the conditions the `debug_unreachable`s are based on?